### PR TITLE
[WIP] [CR] Clothing changes

### PR DESCRIPTION
**Pull request by mugling**
I've begun refactoring the code calculating clothing encumbrance in preparation for adding new features. The function now more closely matches the logic described in the preamble, and some dead code is also removed.

I would like to implement an item flag suppressing the layering penalty for use on small wearable items that currently have no assigned body part (watches, flashlights etc). It should be perfectly possible for zombies to claw off your gold wristwatch or better yet smash your head torch apart leaving you alone in the dark. A large number of other items could also then become wearable (hip flask etc) but with the risk of incurring damage. For balance some items could be marked as FRAGILE or ROBUST? What should the NOLAYERPENALTY flag be named?

Also to implement a penalty for misplacement of layers. Wearing a backpack under a coat is possible and may have its advantages, but would not be especially practical. What would a suitable penalty be and should it be fixed or scale with the number of misplaced layers? Effect should almost certainly apply to small items described above worn at the wrong layer as most of those items would be affixed to your outer layers.

### CODE REVIEW

File: src/player.cpp [hunk 1]
```diff
@@ -12119,51 +12119,26 @@ int player::encumb( body_part bp ) const
...
-            break;
-        }
-    }
+    int layer[MAX_CLOTHING_LAYER] = { };
 
-    for( auto& w : worn ) {
-        if( !w.covers(bp) ) {
-            continue;
-        }
+    for (auto w = worn.begin(); w != worn.end(); ++w) {
```
Comment (kevingranade) [inline src/player.cpp]: Why the change to iterators instead of the range-based loop?

File: src/player.cpp [hunk 2]
```diff
@@ -12744,6 +12719,14 @@ bool player::is_wearing_power_armor(bool *hasHelmet) const {
     return result;
 }
 
+bool player::is_wearing_active_power_armor() const
+{
+    for (auto &w : worn) {
+        if (w.is_power_armor() && w.active) return true;
```
Comment (kevingranade) [inline src/player.cpp]: Style nit, {} are mandatory, and expressions should go on the next line.

File: src/item.cpp [hunk 3]
```diff
@@ -2465,6 +2465,15 @@ int item::get_encumber() const
     return encumber;
 }
 
+int item::get_layer() const
+{
+    if      (has_flag("SKINTIGHT")) return UNDERWEAR;
```
Comment (kevingranade) [inline src/item.cpp]: Styling, {} are mandatory, and return expressions should to on next line.

File: src/player.cpp [hunk 4]
```diff
@@ -12119,51 +12119,26 @@ int player::encumb( body_part bp ) const
...
-            level = WAIST_LAYER;
-        } else if ( w.has_flag( "OUTER" ) ) {
-            level = OUTER_LAYER;
-        } else if ( w.has_flag( "BELTED") ) {
-            level = BELTED_LAYER;
-        } else {
-            level = REGULAR_LAYER;
-        }
+        if (! (w->has_flag("FIT") && w->get_encumber() == 0))
+            layer[w->get_layer()] += 10;
```
Comment (kevingranade) [inline src/player.cpp]: This was intentionally removed, we'd need a good rationale to put it back in, see #12741 for the most recent discussion on the subject, which is leaning toward having the layering penalty scale with underlying encumbrance rather than negating it altogether.

**Comment by karlnp** [comment]
please branch-merge && rebase -i before submitting WIP PRs, because it makes it much easier to tell what you did in Github View

**Comment by kevingranade** [comment]
I'm grabbing the refactors, but the behavioral changes are going in the wrong direction IMO.

**Comment by mugling** [comment]
@karlnp apologies for the botched rebase

File: src/player.cpp (hunk 1 above)
Comment (mugling) [inline src/player.cpp]: i had implemented further functions to test for layer misplacement which would need either an index or iter to find the specific item in the players inventory but this is perhaps redundant for now considering #12741

File: src/player.cpp [hunk 5]
```diff
@@ -12119,51 +12119,25 @@ int player::encumb( body_part bp ) const
...
-        }
-    }
+    int layer[MAX_CLOTHING_LAYER] = { };
 
-    for( auto& w : worn ) {
-        if( !w.covers(bp) ) {
-            continue;
-        }
+    for (auto w = worn.begin(); w != worn.end(); ++w) {
+        if (! w->covers(bp)) continue;
```
Comment (Coolthulhu) [inline src/player.cpp]: Standards:
Each `if` should have braces, even when it is followed only by a single instruction.

File: src/player.cpp [hunk 6]
```diff
@@ -12119,51 +12119,25 @@ int player::encumb( body_part bp ) const
...
         }
     }
     armorenc = std::max(0, armorenc);
-    ret += armorenc;
 
-    for( auto &elem : layer ) {
-        layers += std::max( 0.0, elem - 10.0 );
-    }
+    for (auto &elem : layer)
+       layers += std::max(0, elem - 10);
```
Comment (Coolthulhu) [inline src/player.cpp]: Same for loops

Comment (kevingranade) [inline src/player.cpp]: I fixed up the style issues when I merged the refactors.

File: src/item.cpp [hunk 7]
```diff
@@ -2465,6 +2465,20 @@ int item::get_encumber() const
     return encumber;
 }
 
+int item::get_layer() const
```
Comment (Coolthulhu) [inline src/item.cpp]: You should be returning `layer_level` rather than `int` here.

**Comment by karlnp** [comment]
yeah, instead of merging it'd just be a rebase.
